pkg/cli/agent/agent.go

```diff
@@ -46,6 +47,22 @@ func Run(ctx *cli.Context) error {
 		return err
 	}
 
+	ctx := signals.SetupSignalContext()
+	wg := &sync.WaitGroup{}
+
+	// If exiting due to an error, ensure that contexts are cancelled so that the
+	// WaitGroup exits.  Otherwise, wait for something else to initiate shutdown.
+	defer func() {
+		if rerr != nil {
+			// do not need to pass the error in here, it will be reported by the CLI error handler
+			signals.RequestShutdown(nil)
+		} else {
+			<-ctx.Done()
+			rerr = ctx.Err()
+		}
+		wg.Wait()
```

Comment:
Assuming this will block the K3s process if one goroutine does not properly handle the cancellation of the context, would it perhaps make sense to add a log or something to understand why the K3s process is hanging?

Reply:
Both kine and etcd tell you when they are waiting to shut down. There is nothing else that blocks the WG, so I think logging here would just be additional noise. If we add additional things that block the WG and find that they are holding things up without any feedback, we can address that later.